lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,684 @@
+#include "swift/ClangImporter/BuiltinMappedTypes.def"
+
+    // On 64-bit Windows, no C type is imported as an Int or UInt; CLong is
+    // imported as an Int32 and CLongLong as an Int64. Therefore, manually
+    // add mappings to C for Int and UInt.
+    if (Triple.isOSWindows() && Triple.isArch64Bit()) {
+      // Map UInt to uintptr_t
+      auto swiftUIntType = getNamedSwiftType(stdlib, "UInt");
+      auto clangUIntPtrType = ctx.getCanonicalType(ctx.getUIntPtrType());
+      Cache.insert({swiftUIntType, clangUIntPtrType});
+
+      // Map Int to intptr_t
+      auto swiftIntType = getNamedSwiftType(stdlib, "Int");
+      auto clangIntPtrType = ctx.getCanonicalType(ctx.getIntPtrType());
+      Cache.insert({swiftIntType, clangIntPtrType});
+    }
+    StdlibTypesAreCached = true;
+  }
+
+  auto it = Cache.find(Type(type));
+  if (it != Cache.end())
+    return it->second;
+
+  it = Cache.find(type->getCanonicalType());
+  if (it != Cache.end()) {
+    Cache.insert({Type(type), it->second});
+    return it->second;
+  }
+
+  return clang::QualType();
+}
+
+clang::QualType ClangTypeConverter::visitTupleType(TupleType *type) {
+  unsigned tupleNumElements = type->getNumElements();
+  if (tupleNumElements == 0)
+    return ClangASTContext.VoidTy;
+
+  Type eltTy = type->getElementType(0);
+  for (unsigned i = 1; i < tupleNumElements; i++) {
+    if (!eltTy->isEqual(type->getElementType(i)))
+      // Only tuples where all element types are equal map to fixed-size
+      // arrays.
+      return clang::QualType();
+  }
+
+  auto clangEltTy = convert(eltTy);
+  if (clangEltTy.isNull())
+    return clang::QualType();
+
+  APInt size(32, tupleNumElements);
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+
+  // Single protocol -> id<Proto>
+  if (proto->isObjC()) {
+    auto &clangCtx = ClangASTContext;
+    clang::IdentifierInfo *name = &clangCtx.Idents.get(proto->getName().get());
+    auto *PDecl = clang::ObjCProtocolDecl::Create(
+                    const_cast<clang::ASTContext &>(clangCtx),
+                    clangCtx.getTranslationUnitDecl(), name,
+                    clang::SourceLocation(), clang::SourceLocation(), nullptr);
+
+    // Attach an objc_runtime_name attribute with the Objective-C name to use
+    // for this protocol.
+    SmallString<64> runtimeNameBuffer;
+    PDecl->addAttr(clang::ObjCRuntimeNameAttr::CreateImplicit(
+                     PDecl->getASTContext(),
+                     proto->getObjCRuntimeName(runtimeNameBuffer)));
+
+    auto clangType  = clangCtx.getObjCObjectType(clangCtx.ObjCBuiltinIdTy,
+                                                 &PDecl, 1);
+    return clangCtx.getObjCObjectPointerType(clangType);
+  }
+
+  return getClangIdType(ClangASTContext);
```

Comment:
This doesn't actually work for a non-ObjC existential, and I don't think they're generally bridgeable, either.

Reply:
Does this mean that the original code had a bug, which just didn’t get triggered?

Reply:
Probably just assuming that this is only ever exercised with specific types.

Reply:
Removing this (and crashing instead) causes a couple of tests to fail:

```
Interpreter/SDK/objc_mangling.swift
sodlib/SwiftObjectNSObject.swift
```

Both of those failures are for %target-run lines. I don't want to spend time into digging deeper here right now. Will leave a TODO.